Zilon.Core/Zilon.Core/Tactics/GameLoop.cs

```diff
@@ -1,87 +1,133 @@
 
-                ProcessActor(actor);
+                state.UpdateCounter();
+
+                if (state.TaskIsExecuting)
+                {
+                    state.Task.Execute();
+                    state.TaskSource.ProcessTaskExecuted(state.Task);
+                }
             }
 
-            _sectorManager.CurrentSector.Update();
+            // удаляем выполненные задачи.
+            foreach (var taskStatePair in taskDict.ToArray())
+            {
+                var state = taskStatePair.Value;
+                if (state.TaskComplete)
+                {
+                    taskDict.Remove(taskStatePair.Key);
+                    state.TaskSource.ProcessTaskComplete(state.Task);
+                    continue;
+                }
 
-            Updated?.Invoke(this, new EventArgs());
+                if (taskStatePair.Key.Person.GetModuleSafe<ISurvivalModule>()?.IsDead == true)
+                {
+                    // Персонажи, у которых есть модуль выживания, могут умереть.
+                    // Мертвые персонажы не выполняют задач.
+                    // Их задачи можно прервать, потому что:
+                    // 1. Возможна ситуация, когда мертвый персонаж всё еще выполнить действие.
+                    // 2. Экономит ресурсы.
+                    taskDict.Remove(taskStatePair.Key);
+                    continue;
+                }
+            }
         }
 
-        private void ProcessActor(IActor actor)
+        private async Task GenerateActorTasksAndPutInDictAsync(IEnumerable<IActor> actors, ConcurrentDictionary<IActor, TaskState> taskDict)
         {
-            foreach (var taskSource in ActorTaskSources)
-            {
-                var actorTasks = taskSource.GetActorTasks(actor);
+            var taskSources = _actorTaskSourceCollector.GetCurrentTaskSources();
 
-                foreach (var actorTask in actorTasks)
+            foreach (var actor in actors)
+            {
+                foreach (var taskSource in taskSources)
                 {
-                    try
+                    if (!taskSource.CanGetTask(actor))
                     {
-                        actorTask.Execute();
+                        continue;
                     }
-                    catch (Exception exception)
+
+                    var actorTask = await taskSource.GetActorTaskAsync(actor).ConfigureAwait(false);
+
+                    var state = new TaskState(actor, actorTask, taskSource);
+                    if (!taskDict.TryAdd(actor, state))
                     {
-                        throw new ActorTaskExecutionException($"Ошибка при работе источника команд {taskSource.GetType().FullName}",
-                            taskSource,
-                            exception);
+                        // Это происходит, когда игрок пытается присвоить новую команду,
+                        // когда старая еще не закончена и не может быть заменена.
+                        throw new Exception();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/kreghek/Zilon_Roguelike/pullRequest?prid=5544638)